lightning-invoice/src/lib.rs

```diff
@@ -573,14 +571,48 @@ impl<D: tb::Bool, H: tb::Bool> InvoiceBuilder<D, H, tb::False> {
 	}
 
 	/// Sets the timestamp to the current UNIX timestamp.
-	pub fn current_timestamp(mut self) -> InvoiceBuilder<D, H, tb::True> {
+	pub fn current_timestamp(mut self) -> InvoiceBuilder<D, H, tb::True, C, S> {
 		let now = PositiveTimestamp::from_system_time(SystemTime::now());
 		self.timestamp = Some(now.expect("for the foreseeable future this shouldn't happen"));
 		self.set_flags()
 	}
 }
 
-impl InvoiceBuilder<tb::True, tb::True, tb::True> {
+impl<D: tb::Bool, H: tb::Bool, T: tb::Bool, S: tb::Bool> InvoiceBuilder<D, H, T, tb::False, S> {
+	/// Sets `min_final_cltv_expiry`.
+	pub fn min_final_cltv_expiry(mut self, min_final_cltv_expiry: u64) -> InvoiceBuilder<D, H, T, tb::True, S> {
+		self.tagged_fields.push(TaggedField::MinFinalCltvExpiry(MinFinalCltvExpiry(min_final_cltv_expiry)));
+		self.set_flags()
+	}
+}
+
+impl<D: tb::Bool, H: tb::Bool, T: tb::Bool, C: tb::Bool> InvoiceBuilder<D, H, T, C, tb::False> {
+	/// Sets the payment secret and relevant features.
```

Comment:
Hmm, we still need to be able to set more features here, no? You may want to set some future features which is probably in `known` but not in the set that is required to understand the invoice here.

Reply:
Do you mean setting a feature as optional? If so, that is an open question that I had. For instance, would we want to have both `basic_mpp_required` and `basic_mpp_optional` methdos? It wasn't entirely clear from the BOLT if either were possible:

> - if the `basic_mpp` feature is offered in the invoice:
>     - MAY pay using Basic multi-part payments.
> - otherwise:
>     - MUST NOT use Basic multi-part payments.

Does "offered" mean the odd bit (optional) was set? Is something similar possible for `payment_secret`? The BOLT seems to imply that if a secret is set then it must be used:

> - if there is a valid `s` field:
>     - MUST use that as `payment_secret`

And what would optional mean here if `basic_mpp` is set as required as there is a dependency requirement?

Reply:
> Do you mean setting a feature as optional?

I meant a feature other than the two below. eg what if we have `option_htlcs_colored_blue`? Some users may really like blue HTLCs some users may not, and LDK may reasonably support receiving both. Thus, we need some way to provide a `Features` object, I think.

> Does "offered" mean the odd bit (optional) was set?

I believe it means either. I think the handling of var_len_onion here is fine.

Reply:
> I meant a feature other than the two below. eg what if we have `option_htlcs_colored_blue`? Some users may really like blue HTLCs some users may not, and LDK may reasonably support receiving both. Thus, we need some way to provide a `Features` object, I think.

Ah, so the additional type parameter `S` is for specifying that a secret can be set at most once. Additionally, it is used to condition whether `basic_mpp` can bet set because that requires the `payment_secret` feature and thus also a secret field set in the invoice.

Other features may have similar type parameters associated with them if they require some other fields set in the invoice. For example, if the hypothetical feature were instead `option_htlcs_colored` with an additional field specifying the color blue, then we would add an `option_htlcs_colored` method taking a color to set in the invoice. It would also set the feature bits as necessary.

Long story short is the purpose of `InvoiceBuilder` is to enforce BOLT 11 semantics at compile time. Thus, it can never result in  a `SemanticsError` only a `CreationError`. Allowing features to be set arbitrarily breaks that contract.

Whether the features are set as required or optional is an orthogonal concern. Since these features are optional in `InvoiceFeatures::known()`, then we'd likely want to set them optional here. I think I had decided on required because the dependency between them and wasn't sure what it would mean for some to be optional and others to be required.

Reply:
Hmmmmm, right, for "known" bits that makes sense. I do wonder how we can support "experimental" feature bits (eg DLC invoices or so), but we really need the ability to set them in `InvoiceFeatures` as well which would imply its a separate thing.

Reply:
For experimental features, I'd imagine there could be a separate method to set these which could also check against any known features.